MyCapytain/common/reference/_base.py

```diff
@@ -0,0 +1,463 @@
+from copy import copy
+from typing import Tuple
+from abc import abstractmethod
+
+
+class BaseCitationSet(Exportable):
+    """ A citation set is a collection of citations that optionnaly can be matched using a .match() function
+
+    :param children: List of Citation
+    :type children: [BaseCitation]
+    """
+
+    def __repr__(self):
+        return "<MyCapytain.common.reference.BaseCitationSet object at %s>" % id(self)
+
+    EXPORT_TO = [Mimetypes.JSON.DTS.Std]
+
+    def __init__(self, children=None):
+        self._children = []
+
+        if children:
+            self.children = children
+
+    @property
+    def children(self):
+        """ Children of a citation
+
+        :rtype: [BaseCitation]
+        """
+        return self._children or []
+
+    @children.setter
+    def children(self, val: list):
+        """ Sets children
+
+        :param val: List of citation children
+        """
+        final_value = []
+        if val is not None:
+            for citation in val:
+                if citation is None:
+                    continue
+                elif not isinstance(citation, (BaseCitation, type(self))):
+                    raise TypeError("Citation children should be Citation")
+                else:
+                    if isinstance(self, BaseCitation):
+                        citation.root = self.root
+                    else:
+                        citation.root = self
+                    final_value.append(citation)
+
+        self._children = final_value
+
+    def add_child(self, child):
+        """ Adds a child to the CitationSet
+
+        :param child: Child citation to add
+        :return:
+        """
+        if isinstance(child, BaseCitation):
+            self._children.append(child)
+
+    def __iter__(self):
+        """ Iteration method
+
+        Loop over the citation children
+
+        :Example:
+            >>>    c = BaseCitationSet(name="line")
+            >>>    b = BaseCitationSet(name="poem", children=[c])
+            >>>    b2 = BaseCitationSet(name="paragraph")
+            >>>    a = BaseCitationSet(name="book", children=[b])
+            >>>    [e for e in a] == [a, b, c, b2],
+
+        """
+        for child in self.children:
+            yield from child
+
+    @abstractmethod
+    def match(self, passageId):
+        """ Given a specific passageId, matches the citation to a specific citation object
+
+        :param passageId: Passage Identifier
+        :return: Citation that matches the passageId
+        :rtype: BaseCitation
+        """
+
+    @property
+    def depth(self):
+        """ Depth of the citation scheme
+
+        .. example:: If we have a Book, Poem, Line system,
+         and the citation we are looking at is Poem, depth is 2
+
+
+        :rtype: int
+        :return: Depth of the citation scheme
+        """
+        if len(self.children):
+            return max([child.depth for child in self.children])
```

Comment:
It might, but this I assume people would make sure they have an hardcoded system for such texts :)

Reply:
Perhaps a warning here might be a good idea, then. Something like, "Not all of the subreferences in this text have the same citation depth"? This would let the user know that they need a hard-coded system for this text.